base/loading.jl

```diff
@@ -2,6 +2,52 @@
 
 # Base.require is the implementation for the `import` statement
 
+# Generic case-sensitive version of isfile
+function isfile_casesensitive_slow(path)
+    isfile(path) || return false
+    dir, filename = splitdir(path)
+    any(readdir(dir) .== filename)
+end
+
+isfile_casesensitive(path) = isfile_casesensitive_slow(path)
+
+@windows_only function isfile_casesensitive(path)
+    isfile(path) || return false
+    longpath(path) == path
+end
+
+@linux_only isfile_casesensitive(path) = isfile(path)
+
```

Comment:
On MacOS, could we open the file to get a file descriptor, then call `[fcntl](https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man2/fcntl.2.html)` with `F_GETPATH`, rather than walking the directory tree?